xapian-maintainer-tools/profiling/profiler.c

```diff
@@ -0,0 +1,190 @@
+/** @file profiler.c
+ * @brief C file implementing IO logging
+ */
+/* Compile the profiler with:
+ * gcc -shared -fPIC -o profiler.so profiler.c -ldl
+ *
+ * To use it with a process, use:
+ * LD_PRELOAD=/path/to/profiler.so /path/to/executable
+ *
+ * Running this library will produce an strace.log file,
+ * which can be fed to strace-analyse to produce a log, by running:
+ * ./strace-analyse strace.log
+ */
+
+#define _GNU_SOURCE
+
+#include <dlfcn.h>
+#include <fcntl.h>
+#include <string.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>
+
+#define MAX_CALL_LEN 5000
+
+// function for logging calls
+
+int reset = 0;
+
+void logcall(const char *call)
+{
+    FILE *file_ptr;
+    if (!reset) {
+	file_ptr = fopen("./strace.log", "w");
+	reset = 1;
+    } else {
+	file_ptr = fopen("./strace.log", "a");
+    }
+    fprintf(file_ptr, "%s", call);
+    fclose(file_ptr);
+}
```

Comment:
You're opening and closing the log file for each call logged, which will slow things down unnecessarily.

Also if the implementation of `fopen()` calls `open()` and we intercept that we'll get an infinite recursion.

Reply:
How do you propose I close the strace file? Using __attribute___((destructor))?

Reply:
There's no need to close it explicitly - when a process exits any files it has open will be closed automatically.